File: app_mvp/src/main/java/com/example/jqorz/test1/mvp/delegate/activity/ActivityDelegateImp.java
package com.example.jqorz.test1.mvp.delegate.activity;


import com.example.jqorz.test1.mvp.MvpView;
import com.example.jqorz.test1.mvp.delegate.BaseDelegateCallback;
import com.example.jqorz.test1.mvp.delegate.MvpInternalDelegate;
import com.example.jqorz.test1.mvp.presenters.MvpPresenter;

/**
 * Created by jqorz on 2017/8/28.
 * 实现了ActivityDelegate，在生命周期里控制Presenter与MvpView：
 * 当Activity执行到各个生命周期时，这里会执行对应的生命周期
 */

public class ActivityDelegateImp<P extends MvpPresenter<V>, V extends MvpView> implements ActivityDelegate {
    private final MvpInternalDelegate<P, V> mvpInternalDelegate;

    //传入BaseDelegateCallback 去控制Presenter
    public ActivityDelegateImp(BaseDelegateCallback<P, V> baseMvpDelegateCallback) {
        if (baseMvpDelegateCallback == null)
            throw new NullPointerException("the baseMvpDelegateCallback in ActivityDelegateImp is null");
        mvpInternalDelegate = new MvpInternalDelegate<>(baseMvpDelegateCallback);

    }

    @Override
    public void onCreate() {
        mvpInternalDelegate.createPresenter();
        mvpInternalDelegate.attachView();

    }

    @Override
    public void onPause() {

    }

    @Override
    public void onResume() {

    }

    @Override
    public void onStop() {

    }

    @Override
    public void onDestroy() {
        mvpInternalDelegate.detachView();
    }
}

File: app_mvp/src/main/java/com/example/jqorz/mvptest/lib_kotlin/delegate/activity/ActivityDelegateImp.kt
package com.jqorz.tallybook.base.mvp.delegate.activity

import com.jqorz.tallybook.base.mvp.MvpView
import com.jqorz.tallybook.base.mvp.delegate.BaseDelegateCallback
import com.jqorz.tallybook.base.mvp.delegate.MvpInternalDelegate
import com.jqorz.tallybook.base.mvp.presenters.MvpPresenter

/**
 * 实现了ActivityDelegate，在生命周期里控制Presenter与MvpView：
 * 当Activity执行到各个生命周期时，这里会执行对应的生命周期
 */
class ActivityDelegateImp<P : MvpPresenter<V>?, V : MvpView>(baseMvpDelegateCallback: BaseDelegateCallback<P, V>?) : ActivityDelegate {
    private val mvpInternalDelegate: MvpInternalDelegate<P, V>
    override fun onCreate() {
        mvpInternalDelegate.createPresenter()
        mvpInternalDelegate.attachView()
    }

    override fun onPause() {}
    override fun onResume() {}
    override fun onStop() {}
    override fun onDestroy() {
        mvpInternalDelegate.detachView()
    }

    //传入BaseDelegateCallback 去控制Presenter
    init {
        if (baseMvpDelegateCallback == null) throw NullPointerException("the baseMvpDelegateCallback in ActivityDelegateImp is null")
        mvpInternalDelegate = MvpInternalDelegate(baseMvpDelegateCallback)
    }
}